### ISSUE TITLE
Telecom Search Is Not Working and Returning False Positives

### ISSUE THREAD

**Issue by granadacoder** [microsoft/fhir-server#811]
**Describe the bug**

When I do a search on Patient with telecom filter, the filter is not considered.

R4 with Sql Server backend.

Here is the basic request syntax sugar.

**https://localhost:44348/Patient/?Patient.telecom.where(system='phone')**

The "self" URL is not showing anything telecom related.
        {
            "relation": "self",
            "url": "https://localhost:44348/Patient"
        }

**To Reproduce**

Setup code to run R4 with Sql Server.

Issue following search-patient request:

**https://localhost:44348/Patient/?Patient.telecom.where(system='phone')**

Using "**use**" or "**value**" does not seem to matter. (other "parts" of the telecom).

**Expected behavior**

So if my syntax-sugar is correct, then I expect the filter to kick in and work of course.

If my syntax-sugar is slightly off, this gets into a deeper question.  But personal opinion :: I actually think if a parameter does not "match", there is a danger of returning too many results.  And non-matching inputs should throw an exception.  
Aka, I ~think I did a filter-by-telecom, but it didn't actually take place.
I guess I could compare my input query to the output "self" query, but that against the spirit of things. #philosphizing!

But I expect to get patients with telecom entries that match

                ],
                "name": [
                    {
                        "family": "NotRealFamilyNameOne",
                        "given": [
                            "FirstNameOne",
                            "jacob",
                            "jingerheimer"
                        ]
                    }
                ],
                "telecom": [
                    {
                        "system": "phone",
                        "value": "555-5555-8970",
                        "use": "home"
                    }
                ]

**Actual behavior**

Telecom not considered...shown via the return "self" URL.

        {
            "relation": "self",
            "url": "https://localhost:44348/Patient"
        }

**Comment by hansenms** [comment]
The search is not really a valid search. You are trying to add some sort of FHIR path to the search string. If you want to find all the patients that have a phone number the search would be:

```
GET //fhirserver/Patient?phone:missing=false
```

The behavior of the self link is correct. It should reflect which search parameters were taken into consideration. It is according to spec, please see https://www.hl7.org/fhir/search.html#errors where it says: "In general, servers SHOULD ignore unknown or unsupported parameters...".

You can argue about whether that is preferable. I think there is argument for returning an error on unsupported search parameters, but that is not the spec.

**Comment by granadacoder** [comment]
So how does one search for a phone number value of

555-5555-8970

?

**Comment by hansenms** [comment]
That would be:

```
GET //fhirserver/Patient?phone=555-555-8970
```

**Comment by granadacoder** [comment]
Ok, thanks.  That ?phone= works.

I'm still working out (in my mind) the intuitiveness of that search.  Since the data involves the telecom object.

...............

What threw me off ... was this file:

https://raw.githubusercontent.com/microsoft/fhir-server/master/src/Microsoft.Health.Fhir.R4.Core/Features/Definition/search-parameters.json

(big file)

But in that file, currently is this code:

`"expression" : "Patient.telecom.where(system='email') | Person.telecom.where(system='email') | Practitioner.telecom.where(system='email') | PractitionerRole.telecom.where(system='email') | RelatedPerson.telecom.where(system='email')",
`
and I misintepretted the "expression" for the URL syntax sugar.

But I agree this bug report was not a correct one.  

Thanks for the followups!

**Comment by granadacoder** [comment]
Yeah, here is a more concrete side by side comparison:

If my patient structure for IDENTIFIER looks like this:


            "resource": {
                "resourceType": "Patient",
                "id": "abc0a731-bba2-4060-b813-12b49d0897df",
                "identifier": [
                    {
                        "system": "ssn",
                        "value": "111-11-1111"
                    },
                    {
                        "system": "myuuidsystem",
                        "value": "5eaaefa224534f4590cda4d89c6cdd3b"
                    }

My url-query looks like this:

https://localhost:44348/Patient/?identifier=myuuidsystem|5eaaefa224534f4590cda4d89c6cdd3b

or

https://localhost:44348/Patient/?identifier=ssn|111-11-1111

so I have the "identifier" in the url, then the (value of the) "system" before the pipe, and the (value of the) "value" after the pipe.





But now with telecom


```

            "resource": {
                "resourceType": "Patient",
                "id": "7ade8294-764d-4179-9a0c-019e9fe7365e",
                "identifier": [
                    {
                        "system": "ssn",
                        "value": "111-11-1111"
                    },
                    {
                        "system": "uuid",
                        "value": "d68ca6ac099e4cb4acb86f1837c5f82b"
                    }
                ],
                "telecom": [
                    {
                        "system": "phone",
                        "value": "293-522-8970",
                        "use": "home"
                    }
                ]
```


If I "mimic" the IDENTIFIER search, i would expect the below to work


https://localhost:44348/Patient/?telecom=phone|293-522-8970


But alas, it does not.


I think this is what go me off course with the FHIR-PATH stuff.

**Comment by hansenms** [comment]
Yeah, FHIR search can be a bit counter intuitive at times. Both `identifier` and `telecom` are token search parameters (https://www.hl7.org/fhir/search.html#token). They consist of a system and a code separated by a `|` symbol, e.g. `system | code`, but for `telecom`, which is of type `ContactPoint`, the spec states "For token parameters on elements of type ContactPoint, uri, or boolean, the presence of the pipe symbol SHALL NOT be used - only the [parameter]=[code] form is allowed". 

Hope this clears it up....or that you are at least (like the rest of us) confused at a slightly higher level.

**Comment by granadacoder** [comment]
Thank you for the documentation and followup.  Yes, the mud is much more clear now!

**Comment by granadacoder** [comment]
So if anyone stumbles upon this question.  I just got out of a meeting where we discussed "Where do the short-cuts come from??"

Here is some "mind-bridge" tips.

If you go to HL7.  For the fhir resource. (Patient in this example).

https://www.hl7.org/fhir/patient.html#search

Under "Search Parameters" you'll see these headers:


Name | Type | Description | Expression
-- | -- | -- | --

and one of those rows has this

**phone** | token | A value in a phone contact | **Patient.telecom.where(system='phone')**
-- | -- | -- | --

So basically, I look at "Name" column and find its value (in this case "phone")......and I think of this as the "short cut" to the full Expression, which is "Patient.telecom.where(system='phone')" in this case.

@hansenms  So that's where I'm getting .. how you said "You are trying to add some sort of FHIR path to the search string."  :). 

But the world makes a little more sense now.

But after the meeting, I was like "I need to go update the question I asked way back".

**Comment by hansenms** [comment]
The thing you call a short cut is a search parameter. Specifically a given search parameter, say `phone` or `name` corresponds to a FHIR path expression, which may in fact search multiple fields on the resource. So check `name` for instance, which will look in both first and last name. So the term short cut is probably not recommended. They are just search parameters. They are defined in terms of FHIR Path expressions.

**Comment by granadacoder** [comment]
Good point.  I changed the syntax of my response.
